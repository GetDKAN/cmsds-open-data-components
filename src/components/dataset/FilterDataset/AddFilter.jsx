import React from 'react'
import PropTypes from 'prop-types'
import FontAwesomePro from '../../common/FontAwesomePro/FontAwesomePro'

const AddFilter = ({ addFilter }) => {
  return (
    <button className='add-filter' aria-label='Add filter' onClick={addFilter}>
      <FontAwesomePro icon='plus-circle' size={16} />
      <span>Add new filter</span>
    </button>
  )
}

AddFilter.propTypes = {
  /**
   * Callback function to set filter
   */
  addFilter: PropTypes.func
}

AddFilter.displayName = 'AddFilter'
export default AddFilter
