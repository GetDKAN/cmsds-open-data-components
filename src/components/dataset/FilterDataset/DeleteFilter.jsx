import React from 'react'
import FontAwesomePro from '../../common/FontAwesomePro/FontAwesomePro'
// import { Button } from '@cmsgov/design-system';
import PropTypes from 'prop-types'

const DeleteButton = ({ deleteFilter, enabled, i }) => {
  const deleteClass = enabled ? 'delete-button' : 'delete-button disabled'

  return (
    <button aria-label='delete filter' className={deleteClass} disabled={!enabled} onClick={() => deleteFilter(i)}>
      <FontAwesomePro icon='trash' size={50} />
    </button>
  )
}


// const DeleteButton = ({ deleteFilter, enabled, i }) => {
//   return (
//     <Button
//       aria-label="Delete filter"
//       variation="ghost"
//       size="small"
//       type="button"
//       className="ds-u-margin-top--6 ds-u-padding-right--0"
//       onClick={() => deleteFilter(i)}
//       disabled={!enabled}
//     >
//       Delete filter
//       <span className="fas fa-trash" aria-hidden="true" />
//     </Button>
//   )
// }


DeleteButton.propTypes = {
  /**
   * Callback function to delete filter.
   * Passes `i` prop value as an argument
   */
  deleteFilter: PropTypes.func,
  /**
   * `true` enables the button.
   * `false` disables the button
   */
  enabled: PropTypes.bool,
  /**
   * Index value used for identifying the
   * correct filter to delete. Passed through
   * deleteFilter callback as an argument
   */
  i: PropTypes.number
}

DeleteButton.displayName = 'DeleteButton'
export default DeleteButton
