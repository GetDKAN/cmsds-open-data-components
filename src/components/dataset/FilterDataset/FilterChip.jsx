import React from 'react'
import PropTypes from 'prop-types'
import FontAwesomePro from '../../common/FontAwesomePro/FontAwesomePro'

/*
  wrapping unicode selectors in span tags cause text decoration underline breaks
  The quick solution appears to return entire jsx terms instead of trying to template it.
*/

const FilterChip = ({ filter, deleteFilter, i }) => {
  const { condition, value, column } = filter

  switch (condition) {
    case '=':
      return (
        <button className='chip-body' onClick={() => deleteFilter(i)} aria-label={`Remove ${column} equals ${value} filter`}>
          <FontAwesomePro icon='filter' size={12} fill='#6F757C' aria-hidden />
          "{column}"
          &#65309;&nbsp;
          {value}
          <FontAwesomePro icon='times' size={12} fill='#6F757C' aria-hidden />
        </button>
      )
    case '<>':
      return (
        <button className='chip-body' onClick={() => deleteFilter(i)} aria-label={`Remove ${column} not equal to ${value} filter`}>
          <FontAwesomePro icon='filter' size={12} fill='#6F757C' aria-hidden />
          "{column}"
          &#8800;&nbsp;
          {value}
          <FontAwesomePro icon='times' size={12} fill='#6F757C' aria-hidden />
        </button>
      )
    case '>':
      return (
        <button className='chip-body' onClick={() => deleteFilter(i)} aria-label={`Remove ${column} greater than ${value} filter`}>
          <FontAwesomePro icon='filter' size={12} fill='#6F757C' aria-hidden />
          "{column}"
          &#62;&nbsp;
          {value}
          <FontAwesomePro icon='times' size={12} fill='#6F757C' aria-hidden />
        </button>
      )
    case '<':
      return (
        <button className='chip-body' onClick={() => deleteFilter(i)} aria-label={`Remove ${column} less than ${value} filter`}>
          <FontAwesomePro icon='filter' size={12} fill='#6F757C' aria-hidden />
          "{column}"
          &#60;&nbsp;
          {value}
          <FontAwesomePro icon='times' size={12} fill='#6F757C' aria-hidden />
        </button>
      )
    case 'LIKE':
      return (
        <button className='chip-body' onClick={() => deleteFilter(i)} aria-label={`Remove ${column} contains ${value.split('%').join('')} filter`}>
          <FontAwesomePro icon='filter' size={12} fill='#6F757C' aria-hidden />
          "{column}"
          &#65315;&nbsp;
          {/* strip the sql */}
          {value.split('%').join('')}
          <FontAwesomePro icon='times' size={12} fill='#6F757C' aria-hidden />
        </button>
      )
    case 'starts with':
      return (
        <button className='chip-body' onClick={() => deleteFilter(i)} aria-label={`Remove ${column} starts with ${value.split('%').join('')} filter`}>
          <FontAwesomePro icon='filter' size={12} fill='#6F757C' aria-hidden />
          "{column}"
          ^&nbsp;
          {/* strip the sql */}
          {value.split('%').join('')}
          <FontAwesomePro icon='times' size={12} fill='#6F757C' aria-hidden />
        </button>
      )
    case 'is_empty':
      return (
        <button className='chip-body' onClick={() => deleteFilter(i)} aria-label={`Remove ${column} is empty filter`}>
          <FontAwesomePro icon='filter' size={12} fill='#6F757C' aria-hidden />
          "{column}"
          Is&nbsp;empty&nbsp;
          <FontAwesomePro icon='times' size={12} fill='#6F757C' aria-hidden />
        </button>
      )
    case 'not_empty':
      return (
        <button className='chip-body' onClick={() => deleteFilter(i)} aria-label={`Remove ${column} not empty filter`}>
          <FontAwesomePro icon='filter' size={12} fill='#6F757C' aria-hidden />
          "{column}"
          Not&nbsp;empty&nbsp;
          <FontAwesomePro icon='times' size={12} fill='#6F757C' aria-hidden />
        </button>
      )
    default:
      return (
        <button className='chip-body' onClick={() => deleteFilter(i)} aria-label={`Remove ${column} contains ${value} filter`}>
          <FontAwesomePro icon='filter' size={12} fill='#6F757C' aria-hidden />
          "{column}"
          &#65315;&nbsp;
          {value}
          <FontAwesomePro icon='times' size={12} fill='#6F757C' aria-hidden />
        </button>
      )
  }
}

FilterChip.propTypes = {
  /**
   * Filter data
   */
  filter: PropTypes.shape({
    column: PropTypes.string,
    condition: PropTypes.oneOf([
      '=',
      '<>',
      '>',
      '<',
      'LIKE',
      'starts with',
      'is_empty',
      'not_empty',
      '' // We need to allow an empty case because there are initially no filters so there is no condition set which throws a type error
    ]),
    value: PropTypes.string
  }),
  /**
   * Callback function to delete filter.
   * Passes `i` prop value as an argument
   */
  deleteFilter: PropTypes.func,
  /**
   * Index value used for identifying the
   * correct filter to delete. Passed through
   * deleteFilter callback as an argument
   */
  i: PropTypes.number
}

FilterChip.displayName = 'FilterChip'
export default FilterChip
