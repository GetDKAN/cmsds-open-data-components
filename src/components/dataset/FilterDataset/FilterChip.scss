@use "../../../scss/modules/modules";
@use "../../../scss/modules/colormap";

.chip-list {
  display: flex;
  flex-direction: row;
  flex-wrap: wrap;
  margin-bottom: 12px;

  button {
    display: flex;
    flex-direction: row;
    align-items: center;
    height: 32px;
    background-color: #fff;
    // color: colormap.$charcoal;
    // border: solid 1px colormap.$charcoal-10;
    border-radius: 20px;
    padding: 8px;
    font-family: Muli;
    font-weight: 300;
    font-size: 12px;
    margin: 8px 8px 8px 0;

    &:focus {
      outline: 2px dotted #84898f;
      outline-offset: 2px;
    }

    &.chip-body {
      &:hover {
        text-decoration: underline;
      }
    }

    &.clear-all {
      color: #0c2499;
      padding: 0 8px;
      border: 0;
      background-color: transparent;

      &:hover {
        text-decoration: underline;
      }
    }

    svg {
      margin: 0px 8px 0px 8px;
    }
  }
}
