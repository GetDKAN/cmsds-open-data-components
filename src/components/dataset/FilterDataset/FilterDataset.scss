@use "../../../scss/modules/modules";
@use "../../../scss/modules/colormap";

// wrapper
.ds-c-dialog-wrap {
  .filter-dialog {
    max-width: 768px;
    padding: 16px 0px 0px 0px;
    font-family: Muli;
    div {
      .ds-c-dialog__header {
        padding: 0px;
        display: flex;
        flex-direction: row;
        padding: 0;
        .ds-c-dialog__heading {
          // header
          font-weight: 800;
          align-self: center;
          padding-left: 24px;
          margin-bottom: 0;
          font-size: 20px;
        }
        .ds-c-dialog__close {
          // header button
          align-self: flex-start;
          height: 32px;
          width: 32px;
          margin-right: 15px;
          margin-bottom: 6px;
          color: colormap.$charcoal;
        }
      }
      // main content
      .ds-c-dialog__body {
        div[id^="dialog"][id$="__content"] {
          div {
            // arbitrary container
            display: flex;
            flex-direction: column;
            .info {
              //padding-top: 24px;
              padding-left: 24px;

              span {
                white-space: pre;
                font-size: 12px;
                font-family: Muli;
                display: inline-block;
              }
            }
            .filter-header {
              height: 40px;
              margin-top: 25px;
              display: flex;
              flex-direction: row;
              flex: 1 1 0;
              width: 100%;
              align-items: center;
              column-gap: 16px;
              background-color: colormap.$background-color;
              border-top: solid 1px colormap.$charcoal-10;
              border-bottom: solid 1px colormap.$charcoal-10;
              padding: 0 24px;

              > .header-item {
                display: flex;
                flex-direction: row;
                width: 30%;
                padding-right: 16px;
                flex-grow: 1;

                h3 {
                  font-size: 14px;
                  font-family: Geometria;
                  font-weight: 800;
                }
              }
            }
            .filter-container {
              min-height: 255px;
              max-height: 500px;
              overflow-y: auto;
              .filter-item {
                display: flex;
                flex-direction: row;
                flex: 1 1 0px;
                padding: 0 24px;
                margin-top: 13px;
                padding-bottom: 12px;
                column-gap: 15px;
                border-bottom: solid 1px colormap.$charcoal-20;
                > select, > input {
                  display: flex;
                  padding-left: 16px;
                  padding-right: 30px;
                  font-size: 14px;
                  font-family: Muli;
                  margin: 0;
                  max-height: 36px;
                  padding-right: 40px;
                  text-overflow: ellipsis;
                  option {
                    height: 50px;
                  }
                }
                .delete-button {
                  border: solid 1px #323a45;
                  border-radius: 100%;
                  background-color: #fff;
                  max-height: 36px;
                  cursor: pointer;
                  svg {
                    path {
                      fill: #323a45;
                    }
                  }
                  &.disabled {
                    border: solid 1px colormap.$charcoal-20;
                    svg {
                      path {
                        fill: colormap.$charcoal-20;
                      }
                    }
                  }
                }
              }
            }
            .add-filter {
              display: flex;
              flex-direction: row;
              align-items: center;
              justify-content: center;
              border: solid 2px #c9cee6;
              border-radius: 25px;
              background-color: #fff;
              margin: 13px 24px 14px 24px;
              &:hover {
                // background-color: #c9cee6; at 50%
                background-color: rgba(201, 206, 230, 0.5);
              }
              .fa-plus-circle {
                path {
                  fill: #3c4fac;
                }
              }
              span {
                color: colormap.$blueberry;
                font-size: 16;
                font-weight: 800;
                margin-left: 8px;
              }
            }
          }
        }
      }
      .ds-c-dialog__actions {
        margin-top: 0;
        .filter-dialog-actions {
          display: flex;
          flex-direction: row;
          justify-content: flex-end;
          border-top: solid 1px colormap.$charcoal-20;
          padding: 12px 24px;
          .disabled {
            background-color: colormap.$charcoal-20;
            &:hover {
              color: #fff;
              border-color: colormap.$charcoal-20;
            }
          }
        }
      }
    }
  }
}
