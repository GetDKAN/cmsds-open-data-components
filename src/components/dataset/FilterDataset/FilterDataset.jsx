import React, { useContext } from 'react'
import PropTypes from 'prop-types'
import { Dialog } from '@cmsgov/design-system'
import AddFilter from './AddFilter'
import FilterItem from './FilterItem'
import './FilterDataset.scss'
import { FilteredDispatch } from '../DatasetResource/FilteredDatasetContext'

const FilterDataset = ({
  lFilters,
  applyFilter,
  addFilter,
  updateFilters,
  deleteFilter,
  resetFilters,
  open,
  onExit
}) => {
  const { initOrder } = useContext(FilteredDispatch)

  const applyAndClose = () => {
    applyFilter(lFilters)
    onExit()
  }

  // we want to disable the button if any of the filters have an empty value
  const isEmpty = (filter) => {
    for (const key in filter) {
      // These filters don't need a value.
      if (filter.condition === 'is_empty' || filter.condition === 'not_empty') {
        continue
      }
      if (!filter[key]) {
        return true
      }
    }
    return false
  }
  // How to know when to disable apply button
  const areAnyEmpty = ((allFilters) => {
    if (!allFilters) {
      return true
    }
    // no filters? no accept button
    if (allFilters.length === 0) {
      return true
    }
    for (let i = 0; i < allFilters.length; i++) {
      if (isEmpty(allFilters[i])) {
        return true
      }
    }
    return false
  })(lFilters)

  let applyButtonStyle = 'btn blueberry'
  areAnyEmpty ? applyButtonStyle = 'btn blueberry disabled' : applyButtonStyle = 'btn blueberry'
  const filterMessage = lFilters ? lFilters.length === 1 ? `Apply ${lFilters.length} Filter` : `Apply ${lFilters.length} Filters` : 'Apply 0 Filters'

  return (
    <div className='filter-dialog'>
      <div className={`ds-c-dialog-wrap${open ? ' open' : ''}`}>
        <Dialog
          isOpen={open}
          heading='Filter dataset'
          className='filter-dialog'
          onExit={onExit}
          ariaCloseLabel='Close filter dataset dialog'
          actions={(
            <div className='filter-dialog-actions'>
              <button
                className='btn'
                onClick={resetFilters}
              >
                Reset
              </button>
              <button
                disabled={areAnyEmpty}
                className={applyButtonStyle}
                key='primary'
                onClick={applyAndClose}
              >
                {filterMessage}
              </button>
            </div>
          )}
        >
          {/* content goes here */}
          <div>
            {/* main start */}
            <div className='info'>
              <span>
                Add filters to only display data that meets your criteria. Filtered results can be downloaded.
              </span>
            </div>
            <div className='filter-header'>
              <div className='header-item'>
                <h3>Column name</h3>
              </div>
              <div className='header-item'>
                <h3>Condition</h3>
              </div>
              <div className='header-item'>
                <h3>Value</h3>
              </div>
              <div style={{ minWidth: 36, height: 36, border: 'solid 1px transparent', borderRadius: '100%' }} />
            </div>
            <div className='filter-container'>
              {/* filters go here. */}
              {(lFilters && lFilters.length > 0) && (
                lFilters.map((filter, i) =>
                  (
                    <FilterItem
                      key={`filter-${i}`}
                      i={i}
                      columns={initOrder}
                      filter={filter}
                      filters={lFilters}
                      updateFilters={updateFilters}
                      deleteFilter={deleteFilter}
                    />
                  )
                )
              )}
            </div>
            <AddFilter addFilter={addFilter} />
          </div>
        </Dialog>
      </div>
    </div>
  )
}

FilterDataset.propTypes = {
  /**
   * Array of filters
   */
  lFilters: PropTypes.arrayOf(PropTypes.shape({
    column: PropTypes.string,
    condition: PropTypes.oneOf([
      '=',
      '<>',
      '>',
      '<',
      'LIKE',
      'starts with',
      'is_empty',
      'not_empty',
      '' // We need to allow an empty case because there are initially no filters so there is no condition set which throws a type error
    ]),
    value: PropTypes.string
  })),
  /**
   * Callback function to add a filter
   */
  addFilter: PropTypes.func,
  /**
   * Callback function to update filters
   */
  updateFilters: PropTypes.func,
  /**
   * Callback function to delete a filter
   */
  deleteFilter: PropTypes.func,
  /**
   * Callback function to reset all filters
   */
  resetFilters: PropTypes.func,
  /**
   * Callback function to apply filter(s)
   */
  applyFilter: PropTypes.func,
  /**
   * Callback funtion for dialog window close event
   */
  onExit: PropTypes.func,
  /**
   * `true` shows the component dialog window
   */
  open: PropTypes.bool
}
FilterDataset.displayName = 'FilterDataset'
export default FilterDataset
