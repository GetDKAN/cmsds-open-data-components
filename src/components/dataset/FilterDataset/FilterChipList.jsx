import React from 'react'
import FilterChip from './FilterChip'
import FontAwesomePro from '../../common/FontAwesomePro/FontAwesomePro'
import PropTypes from 'prop-types'
import './FilterChip.scss'

const FilterChipList = ({
  filters,
  deleteFilter,
  resetFilters,
  hidden,
  resetHiddenColumns,
  reordered,
  resetColumnOrder
}) => {
  const showReset = filters.length > 0 || hidden || reordered
  const applyReset = () => {
    resetHiddenColumns()
    resetColumnOrder()
    resetFilters()
  }
  // lets avoid displaying filters if the only one there is our empty starter filter
  return (
    <div className='chip-list'>
      {filters && filters.map((filter, i) => (
        <FilterChip key={`${i}-${filter.div}`} filter={filter} deleteFilter={deleteFilter} i={i} />
      ))}
      {hidden && (
        <button className='chip-body' onClick={resetHiddenColumns}>
          <FontAwesomePro icon='columns' size={12} fill='#323A45' aria-hidden={true} />
          Columns hidden
          <FontAwesomePro icon='times' size={12} fill='#6F757C' aria-hidden={true} />
        </button>
      )}
      {reordered && (
        <button className='chip-body' onClick={resetColumnOrder}>
          <FontAwesomePro icon='columns' size={12} fill='#323A45' aria-hidden={true} />
          Columns reordered
          <FontAwesomePro icon='times' size={12} fill='#6F757C' aria-hidden={true} />
        </button>
      )}
      {showReset && (
        <button className='clear-all' onClick={applyReset}>Clear all filters</button>
      )}
    </div>
  )
}

FilterChipList.propTypes = {
  /**
   * Array of filters
   */
  filters: PropTypes.arrayOf(PropTypes.shape({
    column: PropTypes.string,
    condition: PropTypes.oneOf([
      '=',
      '<>',
      '>',
      '<',
      'LIKE',
      'starts with',
      'is_empty',
      'not_empty',
      '' // We need to allow an empty case because there are initially no filters so there is no condition set which throws a type error
    ]),
    value: PropTypes.string
  })),
  /**
   * `true` if some columns have been hidden
   * as a result of a filter
   */
  hidden: PropTypes.bool,
  /**
   * `true` if columns have been reordered
   */
  reordered: PropTypes.bool,
  /**
   * Callback function to delete filter
   */
  deleteFilter: PropTypes.func,
  /**
   * Callback function to reset the original column order
   */
  resetColumnOrder: PropTypes.func,
  /**
   * Callback function to reset hidden columns
   */
  resetHiddenColumns: PropTypes.func,
  /**
   * Callback function to remove all applied filters
   */
  resetFilters: PropTypes.func
}
FilterChipList.displayName = 'FilterChipList'
export default FilterChipList
