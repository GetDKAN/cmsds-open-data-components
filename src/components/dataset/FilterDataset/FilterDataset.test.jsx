import { render, screen, fireEvent, act } from '@testing-library/react'
import FilterDataset from './FilterDataset'
import FilteredDatasetResource from '../DatasetResource/FilteredDatasetResource'
import { filteredDatasetResource } from '../../../utilities/data-mocks/data-filteredDatasetResource'
import axios from 'axios'

// Mock axios
vi.mock('axios')

// Mock window.scrollTo
global.scrollTo = vi.fn()

const mockApiResponse = {
  'results': [
    {
      measure_name: 'Average Medicare spending associated with an agency\u0027s home health episodes compared with all home health episodes nationally - agency score',
      measure_date_range: 'January 1, 2018 - December 31, 2018'
    },
    {
      measure_name: 'Average Medicare spending associated with an agency\u0027s home health episodes compared with all home health episodes nationally - count',
      measure_date_range: 'January 1, 2018 - December 31, 2018'
    },
    {
      measure_name: 'For patients with diabetes, how often the home health team got doctor\u0027s orders, gave foot care, and taught patients about foot care',
      measure_date_range: 'January 1, 2018 - December 31, 2018'
    },
    {
      measure_name: 'How often home health patients had new or worsened pressure ulcers',
      measure_date_range: 'January 1, 2018 - December 31, 2018'
    },
    {
      measure_name: 'How often home health patients had to be admitted to the hospital',
      measure_date_range: 'January 1, 2018 - December 31, 2018'
    },
    {
      measure_name: 'How often home health patients, who have had a recent hospital stay, had a preventable hospital readmission within 30 days of discharge from home health',
      measure_date_range: 'January 1, 2016 - December 31, 2018'
    },
    {
      measure_name: 'How often patients got better at bathing',
      measure_date_range: 'January 1, 2018 - December 31, 2018'
    },
    {
      measure_name: 'How often patients got better at getting in and out of bed',
      measure_date_range: 'January 1, 2018 - December 31, 2018'
    },
    {
      measure_name: 'How often patients got better at taking their drugs correctly by mouth',
      measure_date_range: 'January 1, 2018 - December 31, 2018'
    },
    {
      measure_name: 'How often patients got better at walking or moving around',
      measure_date_range: 'January 1, 2018 - December 31, 2018'
    },
    {
      measure_name: 'How often patients receiving home health care needed any urgent, unplanned care in the hospital emergency room - without being admitted to the hospital',
      measure_date_range: 'January 1, 2018 - December 31, 2018'
    },
    {
      measure_name: 'How often patients remained at home within 31 days of being discharged from home health',
      measure_date_range: 'January 1, 2017 - December 31, 2018'
    },
    {
      measure_name: 'How often patients\u0027 breathing improved',
      measure_date_range: 'January 1, 2018 - December 31, 2018'
    },
    {
      measure_name: 'How often patients\u0027 wounds improved or healed after an operation',
      measure_date_range: 'January 1, 2018 - December 31, 2018'
    },
    {
      measure_name: 'How often the home health team began their patients\u0027 care in a timely manner',
      measure_date_range: 'January 1, 2018 - December 31, 2018'
    },
    {
      measure_name: 'How often the home health team checked patients for depression',
      measure_date_range: 'January 1, 2018 - December 31, 2018'
    },
    {
      measure_name: 'How often the home health team checked patients\u0027 medications and got doctor\u0027s orders for medication issues in a timely manner',
      measure_date_range: 'January 1, 2018 - December 31, 2018'
    },
    {
      measure_name: 'How often the home health team checked patients\u0027 risk of falling',
      measure_date_range: 'January 1, 2018 - December 31, 2018'
    },
    {
      measure_name: 'How often the home health team made sure that their patients have received a flu shot for the current flu season',
      measure_date_range: 'January 1, 2018 - December 31, 2018'
    },
    {
      measure_name: 'How often the home health team made sure that their patients have received a pneumococcal vaccine (pneumonia shot)',
      measure_date_range: 'January 1, 2018 - December 31, 2018'
    }
  ],
  count: 21,
  schema: {
    'a4eb46a4-fd19-55b0-baf4-de08c3a70abd': {
      fields: {
        measure_name: {
          type: 'text',
          mysql_type: 'text',
          description: 'Measure Name'
        },
        measure_date_range: {
          type: 'text',
          mysql_type: 'text',
          description: 'Measure Date Range'
        }
      }
    }
  },
  query: {
    keys: true,
    limit: 20,
    offset: 0,
    resources: [
      {
        id: 'a4eb46a4-fd19-55b0-baf4-de08c3a70abd',
        alias: 't'
      }
    ],
    count: true,
    results: true,
    schema: true,
    format: 'json',
    rowIds: false,
    properties: [
      'measure_name',
      'measure_date_range'
    ]
  }
}

const mockApplyFilter = vi.fn()
const mockOnExit = vi.fn()
const componentArgs = {
  lFilters: [
    {
      column: 'measure_name',
      condition: 'LIKE',
      value: '%Average%'
    }
  ],
  open: true,
  addFilter: vi.fn(),
  updateFilters: vi.fn(),
  deleteFilter: vi.fn(),
  resetFilters: vi.fn(),
  applyFilter: mockApplyFilter,
  onExit: mockOnExit
}

const renderComponent = async (args = componentArgs) => render (
  <FilteredDatasetResource
    resource={filteredDatasetResource}
  >
    <FilterDataset {...args} />
  </FilteredDatasetResource>
)

describe('FilterDataset component.', () => {
  beforeEach(() => {
    axios.mockResolvedValueOnce({
      data: {
        ...mockApiResponse
      }
    })
  })

  it('Matches snapshot.', async () => {
    let renderedFilterDataset = null
    
    await act(async () => {
      renderedFilterDataset = await renderComponent()
    })

    expect(renderedFilterDataset.asFragment()).toMatchSnapshot()
  })

  it('Dialog is not open when \'open\' prop is false.', async () => {
    let element = null
    
    await act(async () => {
      element = await renderComponent({
        ...componentArgs,
        open: false
      })
    })

    expect(element.container.querySelector('.filter-dialog')).not.toHaveClass('open')
  })

  it('Apply button is disabled when there are no filters.', async () => {
    await act(async () => {
      await renderComponent({
        ...componentArgs,
        lFilters: []
      })
    })

    expect(screen.getByRole('button', { name: 'Apply 0 Filters' })).toHaveClass('disabled')
  })

  it('Apply button is disabled when filters are undefined.', async () => {
    await act(async () => {
      await renderComponent({
        ...componentArgs,
        lFilters: undefined
      })
    })

    expect(screen.getByRole('button', { name: 'Apply 0 Filters' })).toHaveClass('disabled')
  })

  it('Apply button is disabled when filter values are empty.', async () => {
    await act(async () => {
      await renderComponent({
        ...componentArgs,
        lFilters: [
          {
            column: 'measure_name',
            condition: 'LIKE',
            value: ''
          }
        ]
      })
    })

    expect(screen.getByRole('button', { name: 'Apply 1 Filter' })).toHaveClass('disabled')
  })

  it('Apply button is not disabled when filter filter condition is \'is_empty\'.', async () => {
    await act(async () => {
      await renderComponent({
        ...componentArgs,
        lFilters: [
          {
            column: 'measure_date_range',
            condition: 'is_empty',
            value: ''
          }
        ]
      })
    })

    expect(screen.getByRole('button', { name: 'Apply 1 Filter' })).not.toHaveClass('disabled')
  })

  it('Clicking \'Apply Filters\' button applies the filters and closes the dialog.', async () => {
    await act(async () => {
      await renderComponent({
        ...componentArgs,
        lFilters: [
          {
            column: 'measure_name',
            condition: 'is_empty',
            value: ''
          }
        ]
      })
    })

    fireEvent.click(screen.getByRole('button', { name: 'Apply 1 Filter' }))

    expect(mockApplyFilter).toHaveBeenCalledWith([
      {
        column: 'measure_name',
        condition: 'is_empty',
        value: ''
      }
    ])
    expect(mockOnExit).toHaveBeenCalled()
  })
})
