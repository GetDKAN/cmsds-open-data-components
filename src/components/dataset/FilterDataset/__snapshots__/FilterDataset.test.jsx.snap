// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`FilterDataset component. > Matches snapshot. 1`] = `
<DocumentFragment>
  <div
    class="filter-dialog"
  >
    <div
      class="ds-c-dialog-wrap open"
    >
      <dialog
        aria-labelledby="dialog--1__heading"
        class="ds-c-dialog filter-dialog"
        id="dialog--1"
        open=""
        role="dialog"
        tabindex="-1"
      >
        <div
          class="ds-c-dialog__window"
        >
          <div
            class="ds-c-dialog__header"
          >
            <h2
              class="ds-c-dialog__heading"
              id="dialog--1__heading"
            >
              Filter dataset
            </h2>
            <button
              aria-label="Close filter dataset dialog"
              class="ds-c-close-button ds-c-dialog__close"
              id="dialog--1__close"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="ds-c-icon ds-c-icon--close ds-c-icon--close-thin "
                id="dialog--1__close__icon"
                viewBox="-2 -2 18 18"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M0 13.0332964L13.0332964 0M13.0332964 13.0332964L0 0"
                  fill="none"
                  stroke="currentColor"
                  stroke-linecap="round"
                  stroke-width="2"
                />
              </svg>
            </button>
          </div>
          <div
            class="ds-c-dialog__body"
          >
            <div
              id="dialog--1__content"
            >
              <div>
                <div
                  class="info"
                >
                  <span>
                    Add filters to only display data that meets your criteria. Filtered results can be downloaded.
                  </span>
                </div>
                <div
                  class="filter-header"
                >
                  <div
                    class="header-item"
                  >
                    <h3>
                      Column name
                    </h3>
                  </div>
                  <div
                    class="header-item"
                  >
                    <h3>
                      Condition
                    </h3>
                  </div>
                  <div
                    class="header-item"
                  >
                    <h3>
                      Value
                    </h3>
                  </div>
                  <div
                    style="min-width: 36px; height: 36px; border: 1px solid transparent; border-radius: 100%;"
                  />
                </div>
                <div
                  class="filter-container"
                >
                  <div
                    class="filter-item"
                  >
                    <select
                      aria-label="Select column"
                      class="ds-c-field"
                    >
                      <option
                        disabled=""
                        hidden=""
                        value=""
                      >
                        Select column
                      </option>
                      <option
                        disabled=""
                        value="measure_name"
                      >
                        Measure Name
                      </option>
                      <option
                        value="measure_date_range"
                      >
                        Measure Date Range
                      </option>
                    </select>
                    <select
                      aria-label="Select condition"
                      class="ds-c-field"
                    >
                      <option
                        disabled=""
                        hidden=""
                        value=""
                      >
                        Select condition
                      </option>
                      <option
                        value="="
                      >
                        ＝ Equals
                      </option>
                      <option
                        value="<>"
                      >
                        ≠  Not Equal
                      </option>
                      <option
                        value="LIKE"
                      >
                        Ｃ Contains
                      </option>
                      <option
                        value="starts with"
                      >
                        ^ Starts With
                      </option>
                      <option
                        value="is_empty"
                      >
                        Is empty
                      </option>
                      <option
                        value="not_empty"
                      >
                        Not empty
                      </option>
                    </select>
                    <input
                      aria-label="Enter value"
                      placeholder="Enter value"
                      value="Average"
                    />
                    <button
                      aria-label="delete filter"
                      class="delete-button"
                    >
                      <svg
                        aria-label="trash"
                        class="fa fa-trash"
                        data-icon="trash"
                        data-prefix="fal"
                        focusable="false"
                        height="20"
                        role="img"
                        viewBox="0 0 448 512"
                        width="20"
                        xmlns="http://www.w3.org/2000/svg"
                      >
                        <path
                          class=""
                          d="M440 64H336l-33.6-44.8A48 48 0 0 0 264 0h-80a48 48 0 0 0-38.4 19.2L112 64H8a8 8 0 0 0-8 8v16a8 8 0 0 0 8 8h18.9l33.2 372.3a48 48 0 0 0 47.8 43.7h232.2a48 48 0 0 0 47.8-43.7L421.1 96H440a8 8 0 0 0 8-8V72a8 8 0 0 0-8-8zM171.2 38.4A16.1 16.1 0 0 1 184 32h80a16.1 16.1 0 0 1 12.8 6.4L296 64H152zm184.8 427a15.91 15.91 0 0 1-15.9 14.6H107.9A15.91 15.91 0 0 1 92 465.4L59 96h330z"
                          fill="#323A45"
                        />
                      </svg>
                    </button>
                  </div>
                </div>
                <button
                  aria-label="Add filter"
                  class="add-filter"
                >
                  <svg
                    aria-label="plus circle"
                    class="fa fa-plus-circle"
                    data-icon="plus-circle"
                    data-prefix="fal"
                    focusable="false"
                    height="16"
                    role="img"
                    viewBox="0 0 512 512"
                    width="16"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      class=""
                      d="M384 250v12c0 6.6-5.4 12-12 12h-98v98c0 6.6-5.4 12-12 12h-12c-6.6 0-12-5.4-12-12v-98h-98c-6.6 0-12-5.4-12-12v-12c0-6.6 5.4-12 12-12h98v-98c0-6.6 5.4-12 12-12h12c6.6 0 12 5.4 12 12v98h98c6.6 0 12 5.4 12 12zm120 6c0 137-111 248-248 248S8 393 8 256 119 8 256 8s248 111 248 248zm-32 0c0-119.9-97.3-216-216-216-119.9 0-216 97.3-216 216 0 119.9 97.3 216 216 216 119.9 0 216-97.3 216-216z"
                      fill="#323A45"
                    />
                  </svg>
                  <span>
                    Add new filter
                  </span>
                </button>
              </div>
            </div>
            <div
              class="ds-c-dialog__actions"
            >
              <div
                class="filter-dialog-actions"
              >
                <button
                  class="btn"
                >
                  Reset
                </button>
                <button
                  class="btn blueberry"
                >
                  Apply 1 Filter
                </button>
              </div>
            </div>
          </div>
        </div>
      </dialog>
    </div>
  </div>
</DocumentFragment>
`;
